### PR TITLE
feat(amf): Session Modification changes

### PR DESCRIPTION
**Pull request by ajay-kashyap**
feat(amf):Handling  Session Modification  flow changes  in 5G 

Signed-off-by:Ajay-Kashyap  <ajay.kashyap@wavelabs.ai>
## Summary
Changes added :
1. Handling the network initiated session modification request from sessiond towards AMF
2. In AMF : Encoding the Qos Rules and QoS Flow Descriptors and sending it as part of modification request.
3. In NGAP : Encoding the SessionModify Request and sending towards NGAP
4. Handling the response from GNB & UE
5. Sending the corrosponding response towards Sessiond

## Test Plan

    Tested for following scenarios
      1) When Sessiond sends session modification request, AMF receives it and forwards it towards NGAP and Gnb-UE via 
           NGAP-NAS.
      2) AMF to handle the response sent by the UE-Gnb via NAS-NGAP and forwarding the same to sessiond.
      3) Session modification is tested for QoS addition and deletion.

## Additional Information

   Logs are appended in the comments section

### CODE REVIEW

File: lte/gateway/c/core/oai/tasks/grpc_service/AmfServiceImpl.cpp [hunk 1]
```diff
@@ -117,33 +128,139 @@ Status AmfServiceImpl::SetSmfSessionContext(
...
+              .eval_precedence = qos_rule.qos().priority();
+          if (!fillUpPacketFilterContents(
+                  &ul_tft->packetfilterlist.createnewtft[ul_count_packetfilters]
+                       .packetfiltercontents,
+                  &flow.match())) {
+            OAILOG_ERROR(
+                LOG_UTIL,
+                "The uplink packet filter contents are not formatted correctly."
+                "Canceling qos flow creation request. \n");
+            return Status::CANCELLED;
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/grpc_service/AmfServiceImpl.cpp:251]: ENUM cannot be changed..

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp [hunk 2]
```diff
@@ -518,6 +521,27 @@ int amf_app_handle_uplink_nas_message(amf_app_desc_t* amf_app_desc_p,
...
+
+  qos_flow_list_t *current_pti_flow_list = smf_ctx->get_proc_flow_list();
+
+  //Check if its the same procedure received from sessiond
+  if (pdu_session_resp->procedure_trans_identity == smf_ctx->get_pti()) {
+    // Should not happen but a safety check
+    OAILOG_WARNING(LOG_AMF_APP, "Warning PTI from sessiond \n");
+  }
+
+  //Update the PTI and flow list received form sessiond
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: Space added.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.hpp [hunk 3]
```diff
@@ -270,13 +279,31 @@ typedef struct smf_context_s {
...
+  // set current pti from sessiond
+  void set_pti(uint8_t procedure_trans_identity) {
+    smf_proc_data.pti = procedure_trans_identity;
+  }
+
+  // get proc flow list
+  qos_flow_list_t *get_proc_flow_list() {
+    return &(smf_proc_data.qos_flow_list); 
+  }
+
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_ue_context_and_proc.hpp:303]: Added for better readability..

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+    uint16_t pkt_filter_len = 0;
+    pkt_filter = &new_tft[i];
+
+    // Set the spare direction and id
+    new_qos_rule_pkt_filter.spare = 0x0;
+    new_qos_rule_pkt_filter.pkt_filter_dir = pkt_filter->direction;
+    new_qos_rule_pkt_filter.pkt_filter_id = pkt_filter->identifier;
+    uint16_t flag = TRAFFIC_FLOW_TEMPLATE_IPV4_REMOTE_ADDR_FLAG;
+
+    //Check for the max label
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp]: Space added..

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+    while (flag <= TRAFFIC_FLOW_TEMPLATE_MATCH_ALL_FLAG) {
+      switch (pkt_filter->packetfiltercontents.flags & flag) {
+
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp]: added for readability.

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp [hunk 6]
```diff
@@ -0,0 +1,316 @@
...
+        case TRAFFIC_FLOW_TEMPLATE_MATCH_ALL_FLAG: {
+           //Match all type
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp]: Space added..

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp [hunk 7]
```diff
@@ -0,0 +1,316 @@
...
+          new_qos_rule_pkt_filter.contents[pkt_filter_len] =
+              TRAFFIC_FLOW_TEMPLATE_MATCH_ALL;
+          pkt_filter_len++;
+        } break;
+
+        case TRAFFIC_FLOW_TEMPLATE_IPV4_REMOTE_ADDR_FLAG: {
+           //IPv4 remote address type
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp]: Space added..

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp [hunk 8]
```diff
@@ -0,0 +1,316 @@
...
+          new_qos_rule_pkt_filter.contents[pkt_filter_len] =
+              TRAFFIC_FLOW_TEMPLATE_IPV4_REMOTE_ADDR;
+          pkt_filter_len++;
+
+          //Copy IPV4 address and Mask
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp]: Space added..

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp [hunk 9]
```diff
@@ -0,0 +1,316 @@
...
+
+}
+
+// Convert the modification request from sessiond to modification message
+// for this transaction
+int
+amf_smf_session_api_fill_qos_ie_info(std::shared_ptr<smf_context_t> smf_ctx,
+                                     bstring *authorized_qosrules,
+                                     bstring *qos_flow_descriptors) {
+
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp]: added fir readability

[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp]: Added for better readability ..

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp [hunk 10]
```diff
@@ -0,0 +1,316 @@
...
+  uint8_t qos_rules_msg_buffer[QOS_RULES_MSG_BUF_LEN_MAX];
+  uint16_t qos_rules_msg_buf_len = 0;
+
+  uint8_t qos_flow_desc_buffer[QOS_FLOW_DESC_BUF_LEN_MAX];
+  uint16_t qos_flow_desc_buf_len = 0;
+
+  OAILOG_FUNC_IN(LOG_AMF_APP);
+
+  //Prepare for sending message out to UE/GNB
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp]: added space.

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp [hunk 11]
```diff
@@ -0,0 +1,316 @@
...
+  for (int i = 0; i < pti_flow_list->maxNumOfQosFlows; i++) {
+    qos_flow_setup_request_item *qos_flow_req_item =
+          &(pti_flow_list->item[i].qos_flow_req_item);
+
+    //Preparing QoS Rule Msg
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp]: added space.

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp [hunk 12]
```diff
@@ -0,0 +1,316 @@
...
+                          qos_rules_msg_buffer + qos_rules_msg_buf_len,
+                          QOS_RULES_MSG_BUF_LEN_MAX);
+
+      if (encoded_result < 0) {
+        OAILOG_ERROR(LOG_AMF_APP,
+                     "Qos Rule parameters invalid or un-aligned \n");
+        OAILOG_FUNC_RETURN(LOG_AMF_APP, RETURNerror);
+      }
+      qos_rules_msg_buf_len += encoded_result;
+    } //qos rule creation
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp]: added space.

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp [hunk 13]
```diff
@@ -0,0 +1,316 @@
...
+
+    //Preparing qos flow descriptors
+    if (qos_flow_req_item->qos_flow_descriptor.qos_flow_identifier) {
+      M5GQosFlowDescription flow_des = {};
+      flow_des.numOfParams = 0;
+
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp]: added for better readability ..

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp [hunk 14]
```diff
@@ -0,0 +1,316 @@
...
+    if (qos_flow_req_item->qos_flow_descriptor.qos_flow_identifier) {
+      M5GQosFlowDescription flow_des = {};
+      flow_des.numOfParams = 0;
+      
+      qos_flow_descriptor_t* qos_flow_desc =
+          &qos_flow_req_item->qos_flow_descriptor;
+      flow_des.operationCode = qos_flow_req_item->ul_tft.tftoperationcode << 5;
+      flow_des.qfi = qos_flow_desc->qos_flow_identifier;
+
+      //Set  fiveqi flow descriptor
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp]: added space..

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp [hunk 15]
```diff
@@ -0,0 +1,316 @@
...
+
+      if (encoded_result < 0) {
+        OAILOG_ERROR(LOG_AMF_APP,
+                     "Qos Rule parameters invalid or un-aligned \n");
+        OAILOG_FUNC_RETURN(LOG_AMF_APP, RETURNerror);
+      }
+      qos_rules_msg_buf_len += encoded_result;
+    } //qos rule creation
+
+    //Preparing qos flow descriptors
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp]: added space..

File: lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp [hunk 16]
```diff
@@ -0,0 +1,316 @@
...
+      //Set  fiveqi flow descriptor  
+      if (qos_flow_desc->fiveQi) {
+        flow_des.paramList[flow_des.numOfParams].iei =
+            magma5g::M5GQosFlowParam::param_id_5qi;
+        flow_des.paramList[flow_des.numOfParams].length = sizeof(uint8_t);
+        flow_des.paramList[flow_des.numOfParams].element =
+            qos_flow_desc->fiveQi;
+        flow_des.numOfParams++;
+      }
+
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_smf_session_qos.cpp]: for better readability..

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp [hunk 17]
```diff
@@ -883,35 +916,23 @@ int amf_app_handle_pdu_session_accept(
...
-  memcpy(smf_msg->msg.pdu_session_estab_accept.qos_rules.qos_rule, &qos_rule,
-         1 * sizeof(QOSRule));
+  //Check if a default qos rule is received
+  amf_smf_session_set_default_qos_info(smf_ctx);
+
+  amf_smf_session_api_fill_qos_ie_info(smf_ctx,
+     &(smf_msg->msg.pdu_session_estab_accept.authorized_qosrules),
+     &(smf_msg->msg.pdu_session_estab_accept.authorized_qosflowdescriptors));
+
+  //Add length of qos rule
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: Added

File: lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp [hunk 18]
```diff
@@ -883,35 +916,23 @@ int amf_app_handle_pdu_session_accept(
...
+  buf_len += blength(smf_msg->msg.pdu_session_estab_accept
+                     .authorized_qosrules);
+
+  //Add length of qos flow descriptor
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_handler.cpp]: Space included..

File: lte/gateway/c/core/oai/tasks/ngap/ngap_amf_nas_procedures.c [hunk 19]
```diff
@@ -1287,6 +1575,159 @@ int ngap_amf_nas_pdusession_resource_setup_stream(
...
+                     ngap_pdusession_modify_item_ies);
+
+    free(buffer);
+    bdestroy(transfer);
+
+    ASN_STRUCT_FREE_CONTENTS_ONLY(
+        asn_DEF_Ngap_PDUSessionResourceModifyRequestTransfer,
+        pduSessionResourceModifyRequestTransferIEs);
+    free(pduSessionResourceModifyRequestTransferIEs);
+
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/ngap/ngap_amf_nas_procedures.c:1643]: included for readability..

**Comment by rsarwad** [comment]
Could you please update the summary with change description.

**Comment by ajay-kashyap** [comment]
Session Modification flows for addition and deletion of QOS 

image

**Comment by ajay-kashyap** [comment]
NAS message - Addition flow Session Modification 

image

**Comment by ajay-kashyap** [comment]
NGAP- Session modification Addition flow 

image

**Comment by ajay-kashyap** [comment]
NAS PDU Session Modification Deletion Flow 
image

**Comment by ajay-kashyap** [comment]
NGAP Session Modification -Deletion flow
image

**Comment by ajay-kashyap** [comment]
MME Logs 

[mme_session_modification_Add-Delete-flows.log](https://github.com/magma/magma/files/8710172/mme_session_modification_Add-Delete-flows.log)

File: lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp [hunk 20]
```diff
@@ -449,6 +538,77 @@ int send_pdu_resource_setup_response(amf_ue_ngap_id_t ue_id) {
...
+  // traffic flow template
+  response->qos_flow_list.item[0].qos_flow_req_item.ul_tft.tftoperationcode =
+      TRAFFIC_FLOW_TEMPLATE_OPCODE_DELETE_EXISTING_TFT;
+  response->qos_flow_list.item[0]
+      .qos_flow_req_item.ul_tft.numberofpacketfilters = 1;
+
+  // rule id
+  strncpy(reinterpret_cast<char*>(
+             response->qos_flow_list.item[0].qos_flow_req_item.rule_id),
+         rule2,sizeof(rule2));
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp]: Space included.

File: lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp [hunk 21]
```diff
@@ -475,6 +634,94 @@ int send_pdu_notification_response() {
...
+  new_tft->packetfiltercontents.ipv4remoteaddr[1].addr = 10;
+  new_tft->packetfiltercontents.ipv4remoteaddr[2].addr = 2;
+  new_tft->packetfiltercontents.ipv4remoteaddr[3].addr = 2;
+  new_tft->packetfiltercontents.ipv4remoteaddr[0].mask = 0xff;
+  new_tft->packetfiltercontents.ipv4remoteaddr[1].mask = 0xff;
+  new_tft->packetfiltercontents.ipv4remoteaddr[2].mask = 0xff;
+  new_tft->packetfiltercontents.ipv4remoteaddr[3].mask = 0xff;
+  new_tft->packetfiltercontents.singlelocalport = 22334;
+  // rule id
+  strcpy(reinterpret_cast<char*>(
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp]: we have used strncpy

**Review by nstng — APPROVED** [review]
From a Bazel persective LGTM (this includes only the BUILD.bazel change)

**Comment by nstng** [comment]
Is this an updated version of the draft https://github.com/magma/magma/pull/12619 - can the draft be closed?

File: lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp [hunk 22]
```diff
@@ -449,6 +538,76 @@ int send_pdu_resource_setup_response(amf_ue_ngap_id_t ue_id) {
...
+      100000;
+
+  // traffic flow template
+  response->qos_flow_list.item[0].qos_flow_req_item.ul_tft.tftoperationcode =
+      TRAFFIC_FLOW_TEMPLATE_OPCODE_DELETE_EXISTING_TFT;
+  response->qos_flow_list.item[0]
+      .qos_flow_req_item.ul_tft.numberofpacketfilters = 1;
+
+  // rule id
+  strcpy(reinterpret_cast<char*>(
```
[the remark this answers is not in the document]
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp]: replaced with strncpy

File: lte/gateway/c/core/oai/test/ngap/util_ngap_amf_nas_procedures.cpp [hunk 23]
```diff
@@ -207,6 +207,7 @@ bool generator_ngap_pdusession_resource_setup_req(bstring& stream) {
   return true;
 }
 
+#if 0 /* TODO : CLEAN UP NOT NEEDED */
```
Comment (maxhbr) [inline lte/gateway/c/core/oai/test/ngap/util_ngap_amf_nas_procedures.cpp]: still a todo left in the code

File: lte/gateway/c/core/oai/include/n11_messages_types.h [hunk 24]
```diff
@@ -13,7 +13,12 @@
 
 #pragma once
 
+#define MAX_QOS_FLOW 64
+#define MAX_NO_OF_MULTI_CONN 4
+
 #include "lte/gateway/c/core/oai/common/common_types.h"
+// #include "lte/gateway/c/core/oai/lib/3gpp/3gpp_38.413.h"
```
Comment (maxhbr) [inline lte/gateway/c/core/oai/include/n11_messages_types.h]: commented code?

File: lte/gateway/c/core/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp [hunk 25]
```diff
@@ -145,6 +144,84 @@ int pdu_session_resource_setup_request(
...
+  itti_ngap_pdu_session_resource_modify_request_t* ngap_pdu_ses_mod_req =
+      nullptr;
+  qos_flow_add_or_modify_request_list_t* qos_Flow_Add_Or_Modify_List = nullptr;
+  MessageDef* message_p = nullptr;
+
+  message_p = itti_alloc_new_message(TASK_AMF_APP,
+                                     NGAP_PDU_SESSION_RESOURCE_MODIFY_REQ);
+  ngap_pdu_ses_mod_req =
+      &message_p->ittiMsg.ngap_pdu_session_resource_modify_req;
+  // memset(
```
Comment (maxhbr) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp]: commented code

File: lte/gateway/c/core/oai/tasks/amf/amf_app_timer_management.cpp [hunk 26]
```diff
@@ -96,18 +96,24 @@ bool amf_pop_pdu_timer_arg(int timer_id, ue_pdu_id_t* arg) {
 //------------------------------------------------------------------------------
 int AmfUeContext::StartPduTimer(size_t msec, timer_repeat_t repeat,
                                 zloop_timer_fn handler, ue_pdu_id_t arg) {
+#if !MME_UNIT_TEST
```
Comment (maxhbr) [inline lte/gateway/c/core/oai/tasks/amf/amf_app_timer_management.cpp]: test related branching in productive code looks like a bad practice

File: lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp [hunk 27]
```diff
@@ -271,6 +272,7 @@ int send_uplink_nas_registration_complete(amf_app_desc_t* amf_app_desc_p,
   int rc = RETURNerror;
   rc = amf_app_handle_uplink_nas_message(
       amf_app_desc_p, ue_registration_complete, ue_id, originating_tai);
+  // free(ue_registration_complete->data);
```
Comment (maxhbr) [inline lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp]: commented code?

**Review by maxhbr — COMMENTED** [review]
pointing out some bad practices and unclean code

File: lte/gateway/c/core/oai/include/n11_messages_types.h (hunk 24 above)
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/include/n11_messages_types.h]: removed.

File: lte/gateway/c/core/oai/test/ngap/util_ngap_amf_nas_procedures.cpp (hunk 23 above)
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/test/ngap/util_ngap_amf_nas_procedures.cpp]: cleaned up..

File: lte/gateway/c/core/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp (hunk 25 above)
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_pdu_resource_setup_req_rel.cpp]: removed.

File: lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp (hunk 27 above)
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/test/amf/amf_app_test_util.cpp]: removed.

File: lte/gateway/c/core/oai/tasks/amf/amf_app_timer_management.cpp (hunk 26 above)
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/amf/amf_app_timer_management.cpp]: This framework was introduced to keep timers outside the unit testing infrastructure, These are reused from existing 4G framework.

File: lte/gateway/c/core/oai/include/n11_messages_types.h [hunk 28]
```diff
@@ -56,16 +60,59 @@ typedef struct qos_flow_level_qos_parameters_s {
...
+} qos_flow_descriptor_t;
+
 typedef struct qos_flow_setup_request_item_s {
   uint32_t qos_flow_identifier;
   qos_flow_level_qos_parameters qos_flow_level_qos_param;
+  qos_flow_action_t qos_flow_action;
+  qos_flow_descriptor_t qos_flow_descriptor;
+  traffic_flow_template_t ul_tft;
+  uint32_t qos_flow_version;
+#define RULE_ID_LEN 256
```
Comment (ajahl) [inline lte/gateway/c/core/oai/include/n11_messages_types.h]: Can you move the `#define` definition to the top?

**Comment by ajahl** [comment]
Can you arrange the c/c++ Includes as described in https://github.com/magma/magma/issues/12441?

File: lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GQOSRules.hpp [hunk 29]
```diff
@@ -22,20 +22,28 @@ class NewQOSRulePktFilter {
...
 
 // QOSRule Class
 class QOSRule {
  public:
+#define QOS_ADD_RULE_MIN_LEN 3
+#define QOS_DEL_RULE_MIN_LEN 1
   uint8_t qos_rule_id;
   uint16_t len;
   uint8_t rule_oper_code : 3;
+#define QOS_RULE_DQR_BIT_SET 0x1
```
Comment (ajahl) [inline lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GQOSRules.hpp]: Can you move it up to the other 2 #defines?

Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/tasks/nas5g/include/ies/M5GQOSRules.hpp]: its been moved..

File: lte/gateway/c/core/oai/include/n11_messages_types.h (hunk 28 above)
Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/include/n11_messages_types.h]: its been moved..

**Review by ajahl — APPROVED** [review]

**Review by ganeshg87 — APPROVED** [review]
lgtm

File: lte/gateway/c/core/oai/tasks/amf/prepare_request_for_smf.cpp [hunk 30]
```diff
@@ -85,15 +86,92 @@ int create_session_grpc_req_on_gnb_setup_rsp(amf_smf_establish_t* message,
...
   req_rat_specific->set_request_type(magma::lte::RequestType::INITIAL_REQUEST);
-
   TeidSet* gnode_endpoint = req_rat_specific->mutable_gnode_endpoint();
-
   gnode_endpoint->set_teid(message->gnb_gtp_teid);
+  char ipv4_str[INET_ADDRSTRLEN] = {0};
+  inet_ntop(AF_INET, message->gnb_gtp_teid_ip_addr, ipv4_str, INET_ADDRSTRLEN);
+  req_rat_specific->mutable_gnode_endpoint()->set_end_ipv4_addr(ipv4_str);
+
+#if 0 /* TODO : CHECK IT */
```
Comment (panyogesh) [inline lte/gateway/c/core/oai/tasks/amf/prepare_request_for_smf.cpp]: Please remove the todo

**Review by panyogesh — COMMENTED** [review]
Minor comment.

File: lte/gateway/c/core/oai/test/ngap/util_ngap_amf_nas_procedures.cpp [hunk 31]
```diff
@@ -207,6 +207,7 @@ bool generator_ngap_pdusession_resource_setup_req(bstring& stream) {
   return true;
 }
 
+#if 0 /*  CLEAN UP NOT NEEDED */
```
Comment (maxhbr) [inline lte/gateway/c/core/oai/test/ngap/util_ngap_amf_nas_procedures.cpp]: why is that code just commented out? Wouldn't cleanup be better?

Comment (ajay-kashyap) [reply lte/gateway/c/core/oai/test/ngap/util_ngap_amf_nas_procedures.cpp]: Code cleaned up..

**Review by panyogesh — APPROVED** [review]
LGTM

**Pull request merged**